### PR TITLE
refactor: Move directive gathering to SourceCode

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Refactor

#### What changes did you make? (Give an overview)

Continues the work on implementing language plugins by moving directive gathering into `SourceCode`. The changes here primarily affect the flat config path.

I also renamed parts of directive processing to better reflect the reality that some languages might not use comments for directives and instead use actual syntax, so I renamed things from "comment" to "node". Otherwise, the functionality remains the same for apply disable directives. I believe this should now work for any language, but we'll need to do further testing to verify.

refs #16999

#### Is there anything you'd like reviewers to focus on?

I did need to deviate from the RFC design to make this work, in particular, the method needs to returns problems in addition to the directives in order to maintain our current behavior.

We could remove the need to return problems by moving the check for `disable-next-line` back into the linter, as there is an argument to be made that this should carry across any language.

I think the larger question is if we always want this method to be able to return problems, as there may be other validation errors that come up besides the `disable-next-line` check.

### CODE REVIEW

File: lib/source-code/source-code.js
Comment (aladdin-add) [inline lib/source-code/source-code.js:1038]: To confirm: the previous implementation had `problems.push(... .directiveProblems);` here; its was removed now, was that intentional?

Comment (nzakas) [reply lib/source-code/source-code.js:1038]: Yes. Those problems were created via `createDisableDirectives()` in linter. When I moved the functionality here, that function is no longer accessible. It is run inside of linter: https://github.com/eslint/eslint/pull/18328/files#diff-a4ade4bc7a7214733687d84fbb54b625e464d13be7181caf54f564e5985db980R559

**Review by aladdin-add — APPROVED** [review]
LGTM. Would like another review before merging.

File: lib/linter/apply-disable-directives.js
Comment (mdjermanovic) [inline lib/linter/apply-disable-directives.js:162]:
```suggestion
 * Creates a description of deleting an entire unused disable directive.
```

Typo

**Comment by mdjermanovic** [comment]
> We could remove the need to return problems by moving the check for `disable-next-line` back into the linter, as there is an argument to be made that this should carry across any language.
> 
> I think the larger question is if we always want this method to be able to return problems, as there may be other validation errors that come up besides the `disable-next-line` check.

Just to clarify, did you mean `disable-line` (not `disable-next-line`) here? Because `disable-line` are those for which problems can be generated (when they're multiline).

File: lib/source-code/source-code.js
Comment (mdjermanovic) [inline lib/source-code/source-code.js:1036]: This seems different from what the RFC was proposing. Here, SourceCode returns a string `value` that is later parsed in the Linter as a comma-separated list of rule names.

Per the RFC, `SourceCode#getDisableDirectives()` should be doing this parsing, allowing for different formats in different languages. 

```js
interface SourceCode {
    //...
        getDisableDirectives(): Array<DisableDirective>;
    //...
}

interface DisableDirective {
    type: "enable" | "disable";
    ruleId: string;  // <--

    /**
     * The point from which the directive should be applied.
     */
    from: Location;

    /**
     * Location of the comment.
     */
    loc: LocationRange;
}
```

Is the new design intentional?

Comment (nzakas) [reply lib/source-code/source-code.js:1036]: Yes, I did deviate from the RFC here, just forgot to mention that part of the deviation in the original comment.

My thinking here is that if we limit the types of disable directives to the ones we already support, then `applyDisableDirectives()` will just work as-is for all languages and it dramatically reduces the complexity of implementing this feature for other languages.

**Comment by nzakas** [comment]
> Just to clarify, did you mean `disable-line` (not `disable-next-line`) here? Because `disable-line` are those for which problems can be generated (when they're multiline).

I'm talking about this bit of logic: https://github.com/eslint/eslint/pull/18328/files#diff-b141ff11f426cba0e277c36ead873d9644148373b1e7c36111f627c557670329R1012-R1021

**Comment by mdjermanovic** [comment]
> I did need to deviate from the RFC design to make this work, in particular, the method needs to returns problems in addition to the directives in order to maintain our current behavior.
> 
> We could remove the need to return problems by moving the check for `disable-next-line` back into the linter, as there is an argument to be made that this should carry across any language.
> 
> I think the larger question is if we always want this method to be able to return problems, as there may be other validation errors that come up besides the `disable-next-line` check.

I think it's good to leave it possible for SourceCode implementations to return problems related to gathering disable directives because there might be some language-specific kinds of problems we can't be aware of or some restrictions that the implementation would like to impose in order to enforce language-specific best practices.

As for the specific check  - `disable-line` comment/node itself should not be multiline - I think it's fine to leave it to implementations to decide if they want this check. A drawback is that every implementation that wants to enforce this would need to reimplement the check, but it's a simple small check so not a big deal.

Overall, all the deviations from the RFC design look good to me.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
